scenes/game_elements/props/projectile/components/projectile.gd

```diff
@@ -157,8 +157,20 @@ func add_small_fx() -> void:
 func _on_body_entered(body: Node2D) -> void:
 	add_small_fx()
 	duration_timer.start()
-	if body.owner is FillingBarrel:
-		var filling_barrel: FillingBarrel = body.owner as FillingBarrel
+
+	# Retrieve the actual object. If we hit a HitBox, we want its owner.
+	var target_object: Node = body.owner if body.owner else body
+
+	# Logic for Fragile Vase
+	# Delegate hit handling if the object supports it
+	if target_object.has_method("hit_by_droplet"):
+		target_object.hit_by_droplet(label)
+		queue_free()
+		return
```

Comment:
It is a bit confusing to use a variable that can be both the `body` and the `body_owner`. So I would just check:
```
if body.has_method("hit_by_droplet"):
    # logic for fragile barrel
elif body.owner is FillingBarrel:
    # logic for filling barrel
```

By the way, since FragileVase extends FillingBarrel, are you sure that it doesn't work using owner? Because checking if the object is an instance of a class is more reliable than using `has_method`. Please check if this works:
```
if body.owner is FragileVase:
    var fragile_vase: FragileVase = body.owner as FragileVase
    fragile_vase.hit_by_droplet(label)
    # ....
elif body.owner is FillingBarrel:
    # ...
```